Clean shutdown without logged error

eliben opened the pull request:
Avoid logged "use of closed network connection" error in every test/invocation.

There are two scenarios in which a read error is not a "real error":

* EOF - meaning the client closed the connection. This was already handled.
* Server force-closing the connection when `Stop` in invoked. The best way to detect that is to have an explicit flag in the server for "stopping now". Since this is cross-goroutine, the signal is a channel that gets closed.

polinasok commented on service/dap/server.go:
```diff
@@ -42,6 +42,9 @@ type Server struct {
 	conn net.Conn
+	// stopping is a channel that's closed when the server is Stop()-ed. This can
+	// be used to signal to goroutines run by the server that it's time to quit.
+	stopping chan interface{}
```
rpccommon has "stopChan chan struct{}", which they use to justify an Accept failure. I think it is a good idea to match the name. And you could also use it to suppress Accept failure messages on Ctrl-C, can't you?

eliben replied:
Good point - renamed.

Yes, adding it to the Accept failure too

eliben commented:
Thanks for the review - pushed an update. PTAL

eliben commented:
> Won't this block without the default clause if stopChan has not been closed?

Good catch, thanks. Fixed it. Also went ahead and added a test to exercise this (error in reading messages).

While adding it I noticed that when an error occurs, `serveDAPCodec` stops listening on the connection but doesn't close it. So if the client tries to read from the socket - for example expecting a reply to its incorrect message - it gets stuck forever.

I changed the code to close `s.conn` after `serveDAPCodec` returns, and also removed the `nil` guard in `Stop`, because we're never setting it to `nil` anyway. AFAIK calling `Close` on a connection twice is safe - the second call will complain / return an error because it's already closed, but we're not acting upon `Close` errors anyway.

This is not perfect, but without this tests for errors won't work because we're not paying attention to `disconnectChan` in tests. The server expects to signal `disconnectChan` and then to be `Stop()`-ed, but we don't really do this right now. We could change our test to do this - it would be a somewhat larger change, but maybe it's worthwhile because we want to test the whole `disconnectChan` protocal anyway.

WDYT?

eliben commented:
> I am confused. In my previous PR, you strongly encouraged me to move all the shutdown logic to a single place. We ended up with Stop() being the one function that closes things and having Stop() be called from the main goroutine. This appears to be violating that principle. Also, the next step is to have servreDAPCode launch a goroutine per request, which will be sending on conn. Is the intent to have all those writes fail?

The last push was more for discussion than a concrete solution. I wanted to show what's needed to get clean shutdowns the way the code is currently structured. Right now if the server encounters a bad request it doesn't answer anything to the client, stops paying attention to the socket but doesn't close it. So the client will be stuck forever waiting for the server. This is in tests.

It's true that we should strive to shut things down in a single place, but right now it's hard to do in tests because they don't use `disconnectChan`.

The more correct way to solve this would be to stick to the `disconnectChan` protocol in tests as well. 
I'll be happy to take a stab at it later on.

eliben commented:
> Why take out the nil check?
> You could encounter an error before the connection is set and then calling stop on nil would panic. For example, if you Ctrl-C a server before the client connects.

Undone

eliben commented:
> You could use existing client.UnkownRequest() to trigger an error.

Good idea, done!

eliben commented:
> I think you took out that function in your unittest PR.

Looking it again, I can just use another function anyway. Done

eliben commented:
I pushed a new version of the code with a more robust closing sequence. The test now uses `disconnectChan` to `Stop` the server when the server actually expects it.

I inlined server creation into `runTest` because I needed access to `disconnectChan` outside it. Since it's used in a single place, I think inlining is clear. Care is taken not to call `Stop` twice.

I think this should be cleaner - now the protocol is:

1. Server is `Run`
2. If client requests a shutdown or error occurs, `disconnectChan` is triggered
3. The user/caller calls `Stop` after getting a signal on `disconnectChan`

I updated the comment in `server` to clarify that `disconnectChan` is no longer optional. It turned out pretty easy to set it up in a test - so we'll be testing more of the logic this way.

LMK what you think - or we can discuss this in person.

Finally, the commit history is messy now because I merged upstream master. I'll probably just squish this into a single commit when sending upstream

polinasok commented on service/dap/server.go:
```diff
@@ -53,16 +55,16 @@ type Server struct {
-// Stop() on shutdown.
+// via config and assumes its ownership. config.disconnectChan has to be set;
+// it will be closed by the server when the client requested shutdown. Once
+// disconnectChan is closed, Server.Stop() has to be called.
```
must be called

polinasok commented on service/dap/server.go:
```diff
@@ -72,6 +74,7 @@ func NewServer(config *service.Config) *Server {
 // and kills the target process if it was launched by it.
 func (s *Server) Stop() {
```
Please document that this should now be called only once or there will be a panic.

polinasok commented on service/dap/server.go:
```diff
@@ -53,16 +55,16 @@ type Server struct {
-// and the server is ready to be shut down. The caller must call
-// Stop() on shutdown.
+// via config and assumes its ownership. config.disconnectChan has to be set;
+// it will be closed by the server when the client requested shutdown. Once
```
when the client disconnects or requests shutdown
